Essentials/src/main/java/com/earth2me/essentials/commands/Commandeditsign.java

```diff
@@ -32,45 +34,71 @@ protected void run(final Server server, final User user, final String commandLab
                 } else {
+                    final String[] existingLines = sign.getLines();
                     final int line = Integer.parseInt(args[1]) - 1;
-                    sign.setLine(line, "");
-                    sign.update();
+                    existingLines[line] = "";
+
+                    if (callSignEvent(sign, user.getBase(), existingLines)) {
+                        return;
+                    }
+
                     user.sendMessage(tl("editsignCommandClearLine", line + 1));
                 }
-            } else if (args[0].equalsIgnoreCase("copy") || args[0].equalsIgnoreCase("paste")) {
-                final boolean copy = args[0].equalsIgnoreCase("copy");
-                final String tlPrefix = copy ? "editsignCopy" : "editsignPaste";
+            } else if (args[0].equalsIgnoreCase("copy")) {
                 final int line = args.length == 1 ? -1 : Integer.parseInt(args[1]) - 1;
 
                 if (line == -1) {
                     for (int i = 0; i < 4; i++) {
-                        processSignCopyPaste(user, sign, i, copy);
+                        // We use unformat here to prevent players from copying signs with colors that they do not have permission to use.
+                        user.getSignCopy().set(i, FormatUtil.unformatString(user, "essentials.editsign", sign.getLine(i)));
                     }
-                    user.sendMessage(tl(tlPrefix, commandLabel));
+                    user.sendMessage(tl("editsignCopy", commandLabel));
                 } else {
-                    processSignCopyPaste(user, sign, line, copy);
-                    user.sendMessage(tl(tlPrefix + "Line", line + 1, commandLabel));
+                    // We use unformat here to prevent players from copying signs with colors that they do not have permission to use.
+                    user.getSignCopy().set(line, FormatUtil.unformatString(user, "essentials.editsign", sign.getLine(line)));
+                    user.sendMessage(tl("editsignCopyLine", line + 1, commandLabel));
                 }
 
-                if (!copy) {
-                    sign.update();
+            } else if (args[0].equalsIgnoreCase("paste")) {
+                final int line = args.length == 1 ? -1 : Integer.parseInt(args[1]) - 1;
+
+                final String[] existingLines = sign.getLines();
+                if (line == -1) {
+                    for (int i = 0; i < 4; i++) {
+                        existingLines[i] = FormatUtil.formatString(user, "essentials.editsign", user.getSignCopy().get(i));
+                    }
+                    user.sendMessage(tl("editsignPaste", commandLabel));
+                } else {
+                    existingLines[line] = FormatUtil.formatString(user, "essentials.editsign", user.getSignCopy().get(line));
+                    user.sendMessage(tl("editsignPasteLine", line + 1, commandLabel));
                 }
+
+                callSignEvent(sign, user.getBase(), existingLines);
```

Comment:
Nitpick but "sign pasted" messages will show even if the event was cancelled. I don't know if this is important enough to change, though.

Reply:
Could make `callSignEvent` event throw when cancelled, then catch and show a "sign edit cancelled" message so users don't assume the command is just silently broken.

Separately some admins might want to have a way to force sign edits, but that can come later.